arangod/Replication2/ReplicatedLog/LogFollower.cpp

```diff
@@ -0,0 +1,421 @@
+
+
+  {
+    // Preflight checks - does the leader, log and other stuff match?
+    // This code block should not modify the local state, only check values.
+    if (auto result = appendEntriesPreFlightChecks(self.get(), req); result.has_value()) {
+      return *result;
+    }
+
+    self->_lastRecvMessageId = req.messageId;
+  }
+
+  {
+    // Transactional Code Block
+    // This code removes parts of the log and makes sure that
+    // disk and in memory always agree. We first create the new state in memory
+    // as a copy, then modify the log on disk. This is an atomic operation. If
+    // it fails, we forget the new state. Otherwise we replace the old in memory
+    // state with the new value.
+    auto newInMemoryLog = self->_inMemoryLog.takeSnapshotUpToAndIncluding(req.prevLogEntry.index);
+
+    if (self->_inMemoryLog.getLastIndex() != req.prevLogEntry.index) {
+      auto res = self->_logCore->removeBack(req.prevLogEntry.index + 1);
+      if (!res.ok()) {
+        LOG_CTX("f17b8", ERR, _loggerContext)
+            << "failed to remove log entries after " << req.prevLogEntry.index;
+        return AppendEntriesResult::withPersistenceError(_currentTerm, req.messageId, res);
+      }
+    }
+
+    // commit the deletion in memory
+    static_assert(std::is_nothrow_move_assignable_v<decltype(newInMemoryLog)>);
+    self->_inMemoryLog = std::move(newInMemoryLog);
+  }
+
+  struct WaitForQueueResolve {
+    using QueueGuard = Guarded<WaitForQueue, basics::UnshackledMutex>::mutex_guard_type;
+
+    WaitForQueueResolve(QueueGuard guard, LogIndex commitIndex) noexcept
+        : _guard(std::move(guard)),
+          begin(_guard->begin()),
+          end(_guard->upper_bound(commitIndex)) {}
+
+    QueueGuard _guard;
+    WaitForQueue::iterator begin;
+    WaitForQueue::iterator end;
+  };
+
+  // Allocations
+  auto newInMemoryLog = self->_inMemoryLog.append(_loggerContext, req.entries);
+  auto iter = std::make_unique<InMemoryPersistedLogIterator>(req.entries);
+  auto toBeResolvedPtr = std::make_unique<std::optional<WaitForQueueResolve>>();
+
+  auto* core = self->_logCore.get();
+  static_assert(std::is_nothrow_move_constructible_v<decltype(newInMemoryLog)>);
+  auto commitToMemoryAndResolve =
+      [maybeSelf = std::optional<decltype(self)>(std::move(self)),
+       req = std::move(req), newInMemoryLog = std::move(newInMemoryLog),
+       toBeResolvedPtr = std::move(toBeResolvedPtr)](
+          futures::Try<Result>&& tryRes) mutable -> std::pair<AppendEntriesResult, DeferredAction> {
+    // We have to release the guard after this lambda is finished.
+    // Otherwise it would be released when the lambda is destroyed, which
+    // happens *after* the following thenValue calls have been executed. In
+    // particular the lock is held until the end of the future chain is reached.
+    // This will cause deadlocks.
+    TRI_DEFER({ maybeSelf.reset(); });
+    TRI_ASSERT(maybeSelf.has_value());
+    auto& self = maybeSelf.value();
```

Comment:
How about instead of using an optional we simply create a local guard that takes over ownership from the captured guard?
```suggestion
      [selfGuard = std::move(self), req = std::move(req),
       newInMemoryLog = std::move(newInMemoryLog),
       toBeResolvedPtr = std::move(toBeResolvedPtr)](
          futures::Try<Result>&& tryRes) mutable -> std::pair<AppendEntriesResult, DeferredAction> {
    // We have to release the guard after this lambda is finished.
    // Otherwise it would be released when the lambda is destroyed, which
    // happens *after* the following thenValue calls have been executed. In
    // particular the lock is held until the end of the future chain is reached.
    // This will cause deadlocks.
    decltype(selfGuard) self = std::move(selfGuard);
```

Reply:
Great idea! 😍